File: src/main/java/com/dozycoffee/wms/warehouse/domain/valueobject/LocationCode.java
package com.dozycoffee.wms.warehouse.domain.valueobject;

import com.dozycoffee.wms.warehouse.domain.exception.InvalidLocationCodeException;
import java.util.regex.Pattern;

public record LocationCode(String value) {

    private static final Pattern LOCATION_CODE_PATTERN = Pattern.compile("^[A-Z]-\\d{2}$");

    public LocationCode {
        if (value == null || !LOCATION_CODE_PATTERN.matcher(value).matches()) {
            throw new InvalidLocationCodeException();
        }
    }
}

File: src/main/kotlin/com/dozycoffee/wms/warehouse/domain/valueobject/LocationCode.kt
package com.dozycoffee.wms.warehouse.domain.valueobject

import com.dozycoffee.wms.warehouse.domain.exception.InvalidLocationCodeException

@ConsistentCopyVisibility
data class LocationCode private constructor(val value: String) {
    companion object {
        private val LOCATION_CODE_PATTERN = Regex("^[A-Z]-\\d{2}$")

        fun of(value: String?): LocationCode {
            if (value == null || !LOCATION_CODE_PATTERN.matches(value)) {
                throw InvalidLocationCodeException()
            }
            return LocationCode(value)
        }
    }
}
